### PR TITLE
OEL-1439: Project list pages.

### PR DESCRIPTION
**Pull request by drishu**
### Description

Add search api fields and facets needed for the project content type.

### Change log

- Added:
- Changed:
- Deprecated:
- Removed:
- Fixed:
- Security:

### Commands

```sh
[Insert commands here]

```

### CODE REVIEW

File: modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php [hunk 1]
```diff
@@ -337,4 +470,39 @@ protected function assertElementsTexts(array $expected, array $elements): void {
...
+    ]);
+
+    /** @var \Drupal\emr\Entity\EntityMetaInterface $list_page_entity_meta */
+    $list_page_entity_meta = $list_page->get('emr_entity_metas')->getEntityMeta('oe_list_page');
+    /** @var \Drupal\oe_list_pages\ListPageWrapper $list_page_entity_meta_wrapper */
+    $list_page_entity_meta_wrapper = $list_page_entity_meta->getWrapper();
+    $list_page_entity_meta_wrapper->setSource('node', $bundle);
+    $list_page_entity_meta_wrapper->setConfiguration([
+      'override_exposed_filters' => 1,
+      'exposed_filters' => $exposed_filters,
```
Comment (donquixote) [inline modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php]: I like this method. But let's simplify the `$exposed_filters` parameter.
Let it expect a simple array of strings, and then here we can call `array_combine($exposed_filters, $exposed_filters)`.
```suggestion
      'exposed_filters' => array_combine($exposed_filters, $exposed_filters),
```
See other comments above.

File: modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php [hunk 2]
```diff
@@ -148,66 +139,57 @@ public function testCreateListPages() {
...
-    $page->checkField('emr_plugins_oe_list_page[wrapper][exposed_filters][oelp_oe_sc_event__location]');
-    $page->checkField('emr_plugins_oe_list_page[wrapper][exposed_filters][oelp_oe_sc_event__title]');
-    $page->checkField('emr_plugins_oe_list_page[wrapper][exposed_filters][oelp_oe_sc_event__oe_sc_event_dates]');
-    $page->pressButton('Save');
-
-    $this->drupalGet('node/' . $node->id());
+    $list_page = $this->createListPage('oe_sc_event', [
+      'oelp_oe_sc_event__location' => 'oelp_oe_sc_event__location',
+      'oelp_oe_sc_event__title' => 'oelp_oe_sc_event__title',
+      'oelp_oe_sc_event__oe_sc_event_dates' => 'oelp_oe_sc_event__oe_sc_event_dates',
```
Comment (donquixote) [inline modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php]: Let's simplify the parameter, then here we can do this:
```suggestion
      'oelp_oe_sc_event__location',
      'oelp_oe_sc_event__title',
      'oelp_oe_sc_event__oe_sc_event_dates',
```
See my other comment below.

File: modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php [hunk 3]
```diff
@@ -83,24 +83,15 @@ public function testCreateListPages() {
...
-    $this->drupalGet('node/' . $node->id() . '/edit');
-    $page->checkField('Override default exposed filters');
-    $page->checkField('emr_plugins_oe_list_page[wrapper][exposed_filters][oelp_oe_sc_news__title]');
-    $page->checkField('emr_plugins_oe_list_page[wrapper][exposed_filters][oelp_oe_sc_news__oe_publication_date]');
-    $page->pressButton('Save');
-
-    $this->drupalGet('node/' . $node->id());
+    $list_page = $this->createListPage('oe_sc_news', [
+      'oelp_oe_sc_news__title' => 'oelp_oe_sc_news__title',
+      'oelp_oe_sc_news__oe_publication_date' => 'oelp_oe_sc_news__oe_publication_date',
```
Comment (donquixote) [inline modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php]: Simplify as suggested in other comments below:
```suggestion
      'oelp_oe_sc_news__title',
      'oelp_oe_sc_news__oe_publication_date',
```

File: modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php [hunk 4]
```diff
@@ -83,24 +83,15 @@ public function testCreateListPages() {
...
+    ]);
+    $this->drupalGet($list_page->toUrl());
```
Comment (donquixote) [inline modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php:90]: Actually.. now we no longer use the UI to create these list pages. Is this good?
I agree we don't have to do this each time, but at least once, or not?
Why else do we log in as editor?

File: modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php [hunk 5]
```diff
@@ -337,4 +470,39 @@ protected function assertElementsTexts(array $expected, array $elements): void {
...
+   * @param array $exposed_filters
+   *   Facet machine names linked to the bundle's facet source.
+   *
+   * @return \Drupal\node\NodeInterface
+   *   The list page node created.
+   */
+  protected function createListPage(string $bundle, array $exposed_filters): NodeInterface {
+    $list_page = Node::create([
+      'type' => 'oe_list_page',
+      'title' => 'Results',
```
Comment (donquixote) [inline modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php]: I think 'Results' is confusing as a node title here.
```suggestion
      'title' => 'Example list page',
```

File: modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php [hunk 6]
```diff
@@ -255,43 +237,194 @@ public function testCreateListPages() {
...
+    $filter_end->setValue('lt');
+    $dates[1]->setValue($date_plus_10);
+    $search_button->click();
+    $this->assertResultsTitle('Results', 1);
+    $this->assertResults([
+      'Project ongoing',
+    ]);
+
+    // Filter results by status.
+    $filter_form->findButton('Clear filters')->click();
```
Comment (donquixote) [inline modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php]:
```suggestion
    $filter_form->pressButton('Clear filters');
```
This way the IDE does not need to complain about "could be NULL".

File: modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php [hunk 7]
```diff
@@ -255,43 +237,194 @@ public function testCreateListPages() {
...
+    ]);
+
+    // Assert that the filter form for Events exists.
+    $filter_form = $assert_session->elementExists('css', '#oe-list-pages-facets-form');
+    $filter_status = $filter_form->findField('Status');
+    $filter_type = $filter_form->findField('Project type');
+    $filter_budget = $filter_form->findField('Total budget');
+    $filter_start = $filter_form->findField('Start date');
+    $filter_end = $filter_form->findField('End date');
+    $search_button = $filter_form->find('css', '#edit-submit');
```
Comment (donquixote) [inline modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php:351]: We forgot one assertNotNull():
```suggestion
    $search_button = $filter_form->find('css', '#edit-submit');
    $this->assertNotNull($filter_status);
```
My IDE complains if a variable could be NULL..

File: modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php (hunk 6 above)
Comment (donquixote) [reply modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php]: It also throws an exception if the button is not found, which is good.

**Review by donquixote — CHANGES_REQUESTED** [review]

File: modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php (hunk 4 above)
Comment (donquixote) [reply modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php:90]: Discuss in chat if needed.

File: modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php (hunk 1 above)
Comment (drishu) [reply modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php]: If we will ever want to do something like 
```
      'oelp_oe_sc_event__location' => 'oelp_oe_sc_event__location',
      'oelp_oe_sc_event__title' => 0,
      'oelp_oe_sc_event__oe_sc_event_dates' => 'oelp_oe_sc_event__oe_sc_event_dates',
```
we wont be able to, but for our test seems like an edge case, at least at this point, so ok for now, lets see what the future brings.

File: modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php (hunk 4 above)
Comment (drishu) [reply modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php:90]: I yielded

File: modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php (hunk 1 above)
Comment (donquixote) [reply modules/oe_showcase_list_pages/tests/src/ExistingSite/ListPagesTest.php]: Or we could skip that item in that case.
Although, the reason to include it would be to verify that all the options exist.

But for this we could also pass this instead:
```php
      'oelp_oe_sc_event__location' => TRUE,
      'oelp_oe_sc_event__title' => FALSE,
      'oelp_oe_sc_event__oe_sc_event_dates' => TRUE,
```

**Review by donquixote — APPROVED** [review]

**Review by escuriola — APPROVED** [review]
